### DISCUSSION TITLE
Implementation details of alternate screen buffer using virtual codes

### DISCUSSION THREAD

**Discussion by 2trvl** [microsoft/terminal#15920] [Q&A] [answered]
Hi, when I switch to an alternate screen buffer using virtual code `ESC[?1049h`, STD_OUTPUT_HANDLE remains the same. But the WriteConsole function prints the text to an alternate screen. Do I understand correctly that inside terminal this is not arranged as a series of calls:
`CreateConsoleScreenBuffer`
`SetConsoleScreenBufferSize`
`SetConsoleActiveScreenBuffer`
`SetStdHandle`

And the alternate screen is software-based that isn't handled by underlaying API's. At the same time, the buffer created in Windows API is successfully closes using the ANSI ESC code `ESC[?1049l`.

If it's not, how would I be able to detect from Windows Console API that the screen is switched.

**Comment by j4james** [comment, marked as answer]
I don't think you can detect the alternate buffer status with a Console API call, and if you could, it's probably not something you should rely on. However, you can query that state with a standard `DECRQM` escape sequence. If you output `\x1b[?1049$p`, you should get back one of two responses:

* `\x1b[?1049;1$y` if you're in the alternate buffer
* `\x1b[?1049;2$y` if you're in the main buffer

That said, you should be aware that not all terminals will respond to `DECRQM`, and it has only been supported in Windows Terminal since version 1.17.
